### ISSUE TITLE
restoring segmentation for validation set during training

### ISSUE THREAD

**Issue by ashim95** [pytorch/fairseq#426]
Hi fairseq team,
Thanks for this wonderful repository.

When training a new translation model with bpe segmented data using `train.py` file, I see that there's no flag remove-bpe for validation set (so that validation evaluation is performed after restoring the original segmentation). Am I missing something? Or is there a work-around?

**Comment by myleott** [comment]
During validation we only report the NLL loss over the validation set, which is computed over raw tokens (i.e., subword units when using BPE). If you also want to report BLEU scores during validation, then you would need to additionally do inference/generation during the validation phase. This would be quite a bit slower, but then it’d make sense to add `--valid-remove-bpe`, `--valid-beam`, etc.

**Comment by ashim95** [comment]
Thanks for the reply. 

Yeah, I see that BLEU is not calculated during validation (and that generation step is considerably slower). I think I might add the BLEU evaluation for validation also. 

Also, I see that you have used your own BLEU scorer (and not the moses perl script). You report something called BLEU4. Can you tell how that is different from the standard scorer?

**Comment by myleott** [comment]
It should produce the same output as the Moses BLEU scorer but is written in C and is a bit faster. You can compare this using score.py. BLEU4 refers to the n-gram order (=4).
